{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein & DNA enrichment statistics for deletions and substitutions in GFP\n",
    "### eGFP & GFP8; 4 libraries: -3 bp, -6 bp, -9 bp, 3 bp sub; starting library + sorted into 3 fractions each\n",
    "\n",
    "The analysis functions are located in `../indels/proteins.py`, which draws general functions from `../indels/ind.py`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Previous processing of the libraries\n",
    "\n",
    "Reference sequences:\n",
    "\n",
    "`>eGFP\n",
    "atgagtaaaggagaagaacttttcactggagttgtcccaattcttgttgaattagatggtgatgttaatgggcacaaattttctgtcagtggagagggtgaaggtgatgcaacatacggaaaacttacccttaaatttatttgcactactggaaaactacctgttccatggccaacacttgtcactactctgacgtatggtgttcaatgcttttcccgttatccggatcacatgaaacggcatgactttttcaagagtgccatgcccgaaggttatgtacaggaacgcactatattcttcaaagatgacgggaactacaagacgcgtgccgaggtcaagtttgaaggtgatacccttgttaatcgtatcgagttaaaaggtattgattttaaagaagatggaaacattctcggacacaaactcgagtacaactataactcacacaatgtatacatcatggcagacaaacaaaagaatggaatcaaagttaacttcaaaattcgccacaacattgaagatggatccgttcaactagcagaccattatcaacaaaatactccaattggcgatggccctgtccttttaccagacaaccattacctgtcgacacaatctgccctttcgaaagatcccaacgaaaagcgtgaccacatggtccttcttgagtttgtaactgctgctgggattacacatggcatggatgagctctacaaaaAGCTT`\n",
    "\n",
    "`>GFP8\n",
    "atgagtaaaggagaagaacttttcactggagttgtcccaattcttgttgaattagatggtgatgttaatgggcacaaattttctgtccgtggagagggtgaaggtgatgcaacatacggaaaacttacccttaaatttatttgcactactggaaaactacctgttccatggccaacacttgtcactactctgacgtatggtgttcaatgcttttcccgttatccggatcacatgaaacggcatgactttttcaagagtgccatgcccgaaggttatgtacaggaacgcactatatccttcaaagatgacgggacctacaagacgcgtgccgaggtcaagtttgaaggtgatacccttgttaatcgtatcgagttaaaaggtattgattttaaagaagatggaaacattctcggacacaaactcgagtacaactataactcacacaatgtatacatcacggcagacaaacaaaagaatggaatcaaagctaacttcaaaattcgccacaacattgaagatggatccgttcaactagcagaccattatcaacaaaatactccaattggcgatggccctgtccttttaccagacaaccattacctgtcgacacaatctgtcctttcgaaagatcccaacgaaaagcgtgaccacatggtccttcttgagtttgtaactgctgctgggattacacatggcatggatgagctctacaaaaAGCTT`\n",
    "\n",
    "We start with 4 files for eGFP and 4 for GFP8: N (the baseline, unsorted libraries), H, MM and L (pooled sorted libraries). Each of these contains -3, -6, -9 bp deletions and the +-3 bp substitution library.\n",
    "\n",
    "#### Read processing\n",
    "1. Paired End Assembly of Reads\n",
    "```\n",
    "pear-0.9.10-bin-64/pear-0.9.10-bin-64 -f $forwardReads -r $reverseReads -o $baseName.$activity \\\n",
    "--keep-original --min-overlap 5 --min-assembly-length 0 --quality-threshold 15 --max-uncalled-base 0.01\n",
    "```\n",
    "2. Separately, align assembled (in single-end mode) and unassembled reads (in paired-end mode) reads to reference (eGFP.fa or GFP8.fa). Use default Bowtie2 settings. After sorting the SAM files, this generates the depth files.\n",
    "3. Extract non-WT reads based on `grep`. \n",
    "4. Align these \"interesting reads\" with EMBOSS implementation of `needleall` = Needleman-Wunsch algorithm.\n",
    "   - There are rather slow and were originally chosen to give an accurate alignment. There definitely are faster and likely equally fast ways of doing the alignments. One option to explore is https://academic.oup.com/bioinformatics/article/33/9/1394/2964763, which has a Python wrapper available.\n",
    "   - We're aligning short sequencing reads to a reference sequence that is only an order of magnitude longer. The difference between a local and a global alignment will be minimal. Hence, while a global alignment would be preferred, a local algorithm would be acceptable. Edlib has a global & semi-global option.\n",
    "5. Parse alignments to extract mutations\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform the initialization and imports\n",
    "import sys\n",
    "import pickle\n",
    "import re\n",
    "import os\n",
    "import csv\n",
    "import argparse\n",
    "import math\n",
    "import pprint\n",
    "import random\n",
    "\n",
    "from string import ascii_lowercase\n",
    "from collections import Counter, defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Emboss.Applications import NeedleallCommandline\n",
    "\n",
    "# Demand Python 3.\n",
    "if sys.version_info[0] < 3:\n",
    "    print(\"Python 3 is required, but you are using Python %i.%i.%i\") % (\n",
    "        sys.version_info[0], sys.version_info[1], sys.version_info[2])\n",
    "    sys.exit(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All functions for this project are located in `InDelScanner/indels/*.py` files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve the specific functions from ind and proteins.py\n",
    "indels_path=\"/home/maya/InDelScanner\"  # /PATH/TO/InDelScanner\n",
    "if indels_path not in sys.path:\n",
    "    sys.path.append(indels_path)\n",
    "from indels.ind import trim_read, findEnds, endMatch, findGap, gapAlign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/mnt/c/Users/Maya/Dropbox/GFP_deletions\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question 1: Ideal distributions from clean computationally generated dataset\n",
    "\n",
    "We start by making a 'perfect' library that contains all possible mutations of the types we are interesting. Since this analysis is working towards the 2017 sequencing dataset of GFP deletion and subsitution libraries, this dataset contains one read for a -3, -6, -9 bp deletion at any position within eGFP or GFP8, as well as for every NNN substitution. There are no frameshifts here, but we expect to see some stop codons.\n",
    "\n",
    "1. Generate baseline reads with baseline.py: instantaneous for deletions, 1.9 s for substitutions\n",
    "2. Align with needleall: 22.3 s for one set of deletions, 8:49.06 for GFP8-s3, 10:15.79 for eGFP-s3\n",
    "3. Read in with composition scripts: 35:05.50 for all four on a single thread\n",
    "```\n",
    "~/InDelScanner/indels/composition.py -f . -s 9 -e 6 -o GFP.baseline\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from filtering import classify_dna, classify_protein, is_mutation_consecutive\n",
    "from filtering import translate_variant_type\n",
    "from filtering import get_dna_composition, get_full_protein_composition, get_filtered_protein_composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('GFP.baseline.p', 'rb') as f:\n",
    "    tmp = pickle.load(f)\n",
    "\n",
    "tmp.keys()\n",
    "# this has the keys 'eGFP-9-6' and 'GFP8-9-6' to indicate the number of nucleotides padding the gene\n",
    "ideal = {'eGFP': tmp['eGFP-9-6'], 'GFP8': tmp['GFP8-9-6']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['base_d369', 'base_s3'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ideal['eGFP'].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_count_st, dna_reads_st = get_dna_composition(ideal, 1)\n",
    "protein_count_st, protein_reads_st = get_full_protein_composition(ideal, 1)\n",
    "filt_prot_count_st, filt_prot_reads_st = get_filtered_protein_composition(ideal, 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, look at how many variants we tried to generate (ie. the number of input reads) and how many are actually detectable.\n",
    "\n",
    "While pure substitution mutations can be placed very accurately, correct placement of InDels can be inherently ambiguous depending on the sequence context. InDels show some inherent redundancy, where distinct initial transposition and insertion / deletion events result in identical final sequence. For example, in the original sequence …nnGCTACTnn…, -3 bp deletions starting at position 2 (G---CT) and at position 3 (GC---T) result in the same final sequence: nnGCTnn (the remaining sequence context is abbreviated with n). The Needleman-Wunsch algorithm consistently (though arbitrarily) assigns this sequence to a deletion at position 2, such that any deletions originating from position 3 _cannot_ be directly observed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose which \n",
    "row_index_dna = ['d3', 'd6', 'd9', 'f', 's', 'sd', 'other']\n",
    "fractions_base = ['eGFP.base_d369', 'eGFP.base_s3', 'GFP8.base_d369', 'GFP8.base_s3']\n",
    "\n",
    "id_dna_count = pd.DataFrame(dna_count_st, index=row_index_dna, columns=fractions_base)\n",
    "id_dna_reads = pd.DataFrame(dna_reads_st, index=row_index_dna, columns=fractions_base)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Table A**: Maximal theoretical diversity of -3 bp, -6 bp, -9 bp deletion libraries and the ±3 bp substitution library. Derived from generating all possible in-frame mutations with one read per variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.base_d369</th>\n",
       "      <th>eGFP.base_s3</th>\n",
       "      <th>GFP8.base_d369</th>\n",
       "      <th>GFP8.base_s3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>711</td>\n",
       "      <td>0</td>\n",
       "      <td>711</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>706</td>\n",
       "      <td>0</td>\n",
       "      <td>706</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>705</td>\n",
       "      <td>0</td>\n",
       "      <td>705</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>0</td>\n",
       "      <td>42687</td>\n",
       "      <td>0</td>\n",
       "      <td>42687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.base_d369  eGFP.base_s3  GFP8.base_d369  GFP8.base_s3\n",
       "d3                711             0             711             0\n",
       "d6                706             0             706             0\n",
       "d9                705             0             705             0\n",
       "f                   0             0               0             0\n",
       "s                   0         42687               0         42687\n",
       "sd                  0             0               0             0\n",
       "other               0             0               0             0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_dna_reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of DNA 'reads' in the baseline dataset. We generated one read per position (for deletions) or one read per position + codon combination (for substitutions), therefore this table shows the total number of mutations that we tried to generate - ie. the denominator in diversity calculations. If the above issue with InDel redundancy did not exist, this would be the full theoretical diversity. Because of redundancy, it is only the upper limit.\n",
    "\n",
    "**Table B**: Theoretical diversity of same libraries on DNA level, after diversity compression because of InDel redundancy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.base_d369</th>\n",
       "      <th>eGFP.base_s3</th>\n",
       "      <th>GFP8.base_d369</th>\n",
       "      <th>GFP8.base_s3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>527</td>\n",
       "      <td>0</td>\n",
       "      <td>530</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>500</td>\n",
       "      <td>0</td>\n",
       "      <td>504</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>511</td>\n",
       "      <td>0</td>\n",
       "      <td>513</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>0</td>\n",
       "      <td>33029</td>\n",
       "      <td>0</td>\n",
       "      <td>33024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.base_d369  eGFP.base_s3  GFP8.base_d369  GFP8.base_s3\n",
       "d3                527             0             530             0\n",
       "d6                500             0             504             0\n",
       "d9                511             0             513             0\n",
       "f                   0             0               0             0\n",
       "s                   0         33029               0         33024\n",
       "sd                  0             0               0             0\n",
       "other               0             0               0             0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_dna_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once the generated mutations from previous table pass through the alignment & mutation calling, some groups of mutations in neighbouring positions are assigned to the same position - they are redundant. Hence, the theoretical diversity accessible even for a perfect library (which we generated above), is below the maximum of 1 deletion / bp gene length or 64 substitutions (NNN codon) per bp gene length.\n",
    "\n",
    "With redundancy, the theoretical diversity approximately 520 deletions of each length and 33K distinct  substitutions at the DNA level.\n",
    "\n",
    "Next, divide the theoretical diversity after redundancy in Table B (numerator) by the maximal theoretical diversity in Table A (denominator, previous table) to generate **Table C**: average proportion of variants accessible by InDel mutageneis per bp of gene length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.base_d369</th>\n",
       "      <th>eGFP.base_s3</th>\n",
       "      <th>GFP8.base_d369</th>\n",
       "      <th>GFP8.base_s3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>0.741210</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.745429</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>0.708215</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.713881</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>0.724823</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.727660</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.773748</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.773631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.base_d369  eGFP.base_s3  GFP8.base_d369  GFP8.base_s3\n",
       "d3           0.741210           NaN        0.745429           NaN\n",
       "d6           0.708215           NaN        0.713881           NaN\n",
       "d9           0.724823           NaN        0.727660           NaN\n",
       "f                 NaN           NaN             NaN           NaN\n",
       "s                 NaN      0.773748             NaN      0.773631\n",
       "sd                NaN           NaN             NaN           NaN\n",
       "other             NaN           NaN             NaN           NaN"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_dna_count.div(id_dna_reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dividing the two dataframes, we see the theoretical diversity is ~0.74 deletions / bp gene lenght and 0.774×64=49.5 substitutions / bp gene length.\n",
    "\n",
    "**Table D**: Composition of ideal variants when translated to protein. Due to degeneracy of the codon table, the distinct DNA variants in each library results in an equal or lower diversity once the sequences are translated into protein. Synonymous variants that have different DNA and identical protein sequence may still show a difference in expression levels, so the libraries are worth analysing both on protein and DNA level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.base_d369</th>\n",
       "      <th>eGFP.base_s3</th>\n",
       "      <th>GFP8.base_d369</th>\n",
       "      <th>GFP8.base_s3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>230</td>\n",
       "      <td>0</td>\n",
       "      <td>231</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>174</td>\n",
       "      <td>0</td>\n",
       "      <td>174</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>230</td>\n",
       "      <td>0</td>\n",
       "      <td>230</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>236</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>171</td>\n",
       "      <td>0</td>\n",
       "      <td>172</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>0</td>\n",
       "      <td>4514</td>\n",
       "      <td>0</td>\n",
       "      <td>4514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ss</th>\n",
       "      <td>0</td>\n",
       "      <td>4184</td>\n",
       "      <td>0</td>\n",
       "      <td>4145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>25</td>\n",
       "      <td>386</td>\n",
       "      <td>25</td>\n",
       "      <td>383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        eGFP.base_d369  eGFP.base_s3  GFP8.base_d369  GFP8.base_s3\n",
       "d                  230             0             231             0\n",
       "c-sd               174             0             174             0\n",
       "c-dd               230             0             230             0\n",
       "c-sdd              159             0             159             0\n",
       "c-ddd              236             0             236             0\n",
       "c-sddd             171             0             172             0\n",
       "s                    0          4514               0          4514\n",
       "c-ss                 0          4184               0          4145\n",
       "*                   25           386              25           383"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index_prot = ['d', 'c-sd', 'c-dd', 'c-sdd','c-ddd', 'c-sddd', 's', 'c-ss', '*']\n",
    "id_filt_prot = pd.DataFrame(filt_prot_count_st, index=row_index_prot, columns=fractions_base)\n",
    "id_filt_prot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since InDel events (nor substitutions) necessarily occur between codon boundaries, about 40% of mutations of each type introduce an adjacent substitution. Key:\n",
    "- d: pure 1 amino acid deletion\n",
    "- c-sd: 1 aa substitution + 1 aa deletion, consecutive\n",
    "- c-dd: consecutive 2 aa deletion\n",
    "- c-ddd: conscutive 3 aa deletion\n",
    "- s: 1 aa substitution\n",
    "- c-ss: substitutions in 2 consecutive amino acids\n",
    "- `*`: stop codon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question 2: Composition of libraries before sorting\n",
    "\n",
    "The dataset is structured as follows:\n",
    "- eGFP: pooled pre-sorting libraries, in the 'baseline' fraction. This contains the -3 bp, -6 bp, -9 bp TRIAD libraries and the +-3 bp TriNEx substitution library. These were sorted separately, but pooled as one multiplexed fraction during Illumina sequencing.\n",
    "- GFP8: likewise on the stabilized GFP\n",
    "\n",
    "Sorted libraries: Each starting library (4 for eGFP, 4 for GFP8) was sorted into three fractions, separately. After sorting, the equal activity fractions were pooled between variants in the same GFP background and sequenced during the same run as input libraries. This gives 3 fractions (high activity, medium, negative) for each background.\n",
    "\n",
    "The `baseline` fractions give us starting information on the number of variants, their position, proportion of frameshifts - although this will be masked by pooling of libraries made with a different number of cloning steps. It is confounded by sequencing errors, which take the form of rare - usually observed in a single read - single nucleotide substitutions.\n",
    "\n",
    "#### How many variants do we see?\n",
    "#### Where in the parental gene are they located?\n",
    "#### How good is the coverage?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load results, which are in the form of a nested dictionary. The levels are:\n",
    "- results (top level)\n",
    "- protein background: 'eGFP' or 'GFP8'\n",
    "- FACS fractions: 'H', 'L', 'MM', 'baseline'\n",
    "- protein mutations in tuple format: eg. ((204, 's', 'L'), (222, 's', 'G')) or ((130,'d'),)\n",
    "- data for each protein mutation:\n",
    "    - 'total' : total number of reads supporting this protein mutation. This aggregates data from distinct DNA mutations that are synonymous in their effect on protein sequence, including likely sequencing errors.\n",
    "    - 'depth' : average sequencing depth over this mutation\n",
    "    - 'dna' : dictionary listing counts for each DNA mutation, ie. {(DNA tuple) : integer }\n",
    "    - 'dna_hgvs' : a dictionary listing counts for each DNA mutation\n",
    "        - The key for each distinct DNA mutation is the HGVS-like standard for describing DNA variation\n",
    "        - Associated with each key is the count of reads that contain this DNA mutation\n",
    "        \n",
    "To access information for particular protein mutation:\n",
    "\n",
    "results['eGFP']['baseline'][protein_tuple]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['baseline', 'L', 'H', 'MM'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('gfp_with_hgvs_20180827.p', 'rb') as f:\n",
    "    results = pickle.load(f)\n",
    "\n",
    "results['eGFP'].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this stage, we are interested in the pre-sorting situation. Hence, this is an extension of the analysis before to see the distributions in the 'baseline' fractions.\n",
    "\n",
    "**Table E**: The number of all variants observed at least 1× in the unsorted libraries, combined. We see the numbers of detected deletions are  one the same order of magnitude to those in the 'ideal' case, indicating that these are variants in the library. However, the number of substitutions is much inflated: this category on DNA level contains the true substitutions, as well as all the sequencing errors. This is further supported by the large _number_ of detected 'sd' (substitution + deletion) libraries, far exceeding the number of detected deletions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_count, dna_reads = get_dna_composition(results, 1)\n",
    "protein_count, protein_reads = get_full_protein_composition(results, 1)\n",
    "filt_prot_count, filt_prot_reads = get_filtered_protein_composition(results, 1)\n",
    "\n",
    "# Choose which \n",
    "fractions_all = ['eGFP.baseline', 'eGFP.H', 'eGFP.MM', 'eGFP.L', 'GFP8.baseline', 'GFP8.H', 'GFP8.MM', 'GFP8.L']\n",
    "\n",
    "df_dna_count = pd.DataFrame(dna_count, index=row_index_dna, columns=fractions_all)\n",
    "df_dna_reads = pd.DataFrame(dna_reads, index=row_index_dna, columns=fractions_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>443</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>436</td>\n",
       "      <td>439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>394</td>\n",
       "      <td>399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>4304</td>\n",
       "      <td>5687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>17551</td>\n",
       "      <td>19861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>4388</td>\n",
       "      <td>3830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>161</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.baseline  GFP8.baseline\n",
       "d3               443            434\n",
       "d6               436            439\n",
       "d9               394            399\n",
       "f               4304           5687\n",
       "s              17551          19861\n",
       "sd              4388           3830\n",
       "other            161            203"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dna_count[['eGFP.baseline', 'GFP8.baseline']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>40073</td>\n",
       "      <td>29830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>34508</td>\n",
       "      <td>19252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>34776</td>\n",
       "      <td>30195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>23792</td>\n",
       "      <td>39243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>107685</td>\n",
       "      <td>122277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>4800</td>\n",
       "      <td>4184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>1289</td>\n",
       "      <td>1455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.baseline  GFP8.baseline\n",
       "d3             40073          29830\n",
       "d6             34508          19252\n",
       "d9             34776          30195\n",
       "f              23792          39243\n",
       "s             107685         122277\n",
       "sd              4800           4184\n",
       "other           1289           1455"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dna_reads[['eGFP.baseline', 'GFP8.baseline']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that the number of reads corresponding to 'sd' variants is an order of magnitude smaller than reads supporting true deletions, which is in line with the premise that those mutations are primarily sequencing errors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Possible space for improvement: break down the DNA classification to reads containing:**\n",
    "- only a single nucleotide mutation (could be a true variant or sequencing error)\n",
    "- double nucleotide mutations, eg nnAATnn > nnACAnn (generated by TriNEx)\n",
    "- triple nucleotide mutations (genereted by TriNeEx)\n",
    "- non-consecutive single nucleotide mutations (sequencing errors)\n",
    "\n",
    "How to do this? If the classification function on DNA level reads as a substitution, extract the positions and check for conscutiveness. Make sure the NTN/NCN/NAN/NGN mutations are accepted, ie. a triplet exchange where the first and last positions change but the middle is accidentally the same."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Divide this by the ideal distribution calculated before to see what % deletions and substitutions we are seeing. The substitution figure is less reliable.\n",
    "\n",
    "**Table F**: Proportion of observed DNA mutations in unsorted libraries compared to ideal composition - ie. what % of positions do we access."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>0.840607</td>\n",
       "      <td>0.818868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>0.872000</td>\n",
       "      <td>0.871032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>0.771037</td>\n",
       "      <td>0.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>0.531382</td>\n",
       "      <td>0.601411</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.baseline  GFP8.baseline\n",
       "d3          0.840607       0.818868\n",
       "d6          0.872000       0.871032\n",
       "d9          0.771037       0.777778\n",
       "f                inf            inf\n",
       "s           0.531382       0.601411\n",
       "sd               inf            inf\n",
       "other            inf            inf"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_dna_count['eGFP.baseline'] = id_dna_count['eGFP.base_d369'] + id_dna_count['eGFP.base_s3']\n",
    "id_dna_count['GFP8.baseline'] = id_dna_count['GFP8.base_d369'] + id_dna_count['GFP8.base_s3']\n",
    "\n",
    "dna_base_compare = df_dna_count[['eGFP.baseline', 'GFP8.baseline']]\n",
    "dna_base_compare.div(id_dna_count[['eGFP.baseline', 'GFP8.baseline']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Interpretation of starting DNA library composition with non-stringent filtering\n",
    "\n",
    "We see 300-450 variants per starting. The GFP gene is ~720 bp long, depending on where one sets the end of the gene (HindIII restriction site, His tag, barcode etc.). The theoretical diversity of an InDel library depends on the starting sequence: we know from analysis of PTE data that for deletions, the maximum diversity is about 0.75 distinct deletions per bp of gene length.\n",
    "\n",
    "expected max size = 0.75 × 720 = 540 variants\n",
    "\n",
    "The sequencing coverage is very low in the first and last 50 bp of the sequenced fragments, which was a technical limitation of the restriction site positions. Accounting for that, the effective fragment length is ~620 bp and we should see ~465 variants per library for full coverage.\n",
    "\n",
    "We see that the baseline libraries contain 400-440 variants, more for the smaller and less for the larger variants. Hence, the starting detection looks good.\n",
    "\n",
    "There is a substantial number of substitutions (`s`), which are mostly WT reads with sequencing errors, and combinations of deletions + substitutions (`sd`), where a true deletion is accompanied by a sequencing error. There may be true variants hidden in these fractions. The total number of variants with sub substitutions is much larger than pure target variants: for a common variant, there is only one correct variant but many ways it can be accompanied by a  sequencing error. In this way sequencing errors inflate the number of variants, by creating lots of rare variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>40073</td>\n",
       "      <td>29830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>34508</td>\n",
       "      <td>19252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>34776</td>\n",
       "      <td>30195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>23792</td>\n",
       "      <td>39243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>107685</td>\n",
       "      <td>122277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>4800</td>\n",
       "      <td>4184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>1289</td>\n",
       "      <td>1455</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.baseline  GFP8.baseline\n",
       "d3             40073          29830\n",
       "d6             34508          19252\n",
       "d9             34776          30195\n",
       "f              23792          39243\n",
       "s             107685         122277\n",
       "sd              4800           4184\n",
       "other           1289           1455"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dna_reads[['eGFP.baseline', 'GFP8.baseline']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The previous intuition is confirmed here: deletions in each library, each background are supported by 20K-40K reads each. Pure substitutions are still abundant, becaue the analysis starts with selecting ALL reads that deviate from WT - ie. we capture the true variants as well as all the noise.\n",
    "\n",
    "\n",
    "#### Increase filtering stringency to 10 to examine substitutions\n",
    "\n",
    "Re-examine statistics tables with more stringent filtering, to only consider mutations that were observed 5 or more times on DNA level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>0.830700</td>\n",
       "      <td>0.811060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>0.855505</td>\n",
       "      <td>0.790433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>0.829949</td>\n",
       "      <td>0.796992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>0.223745</td>\n",
       "      <td>0.257078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>0.285169</td>\n",
       "      <td>0.269221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>0.002963</td>\n",
       "      <td>0.003133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>0.130435</td>\n",
       "      <td>0.133005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.baseline  GFP8.baseline\n",
       "d3          0.830700       0.811060\n",
       "d6          0.855505       0.790433\n",
       "d9          0.829949       0.796992\n",
       "f           0.223745       0.257078\n",
       "s           0.285169       0.269221\n",
       "sd          0.002963       0.003133\n",
       "other       0.130435       0.133005"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_count_r, dna_reads_r = get_dna_composition(results, 5)\n",
    "df = pd.DataFrame(dna_count_r, index=row_index_dna, columns=fractions_all)\n",
    "df[['eGFP.baseline', 'GFP8.baseline']].div(df_dna_count[['eGFP.baseline', 'GFP8.baseline']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d3</th>\n",
       "      <td>0.996157</td>\n",
       "      <td>0.994335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d6</th>\n",
       "      <td>0.996146</td>\n",
       "      <td>0.988313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d9</th>\n",
       "      <td>0.996003</td>\n",
       "      <td>0.993973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>0.783246</td>\n",
       "      <td>0.835665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>0.812852</td>\n",
       "      <td>0.811469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sd</th>\n",
       "      <td>0.025417</td>\n",
       "      <td>0.022706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>0.856478</td>\n",
       "      <td>0.834364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       eGFP.baseline  GFP8.baseline\n",
       "d3          0.996157       0.994335\n",
       "d6          0.996146       0.988313\n",
       "d9          0.996003       0.993973\n",
       "f           0.783246       0.835665\n",
       "s           0.812852       0.811469\n",
       "sd          0.025417       0.022706\n",
       "other       0.856478       0.834364"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(dna_reads_10, index=row_index_dna, columns=fractions_all)\n",
    "df[['eGFP.baseline', 'GFP8.baseline']].div(df_dna_reads[['eGFP.baseline', 'GFP8.baseline']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above tables show that by filtering the composition analysis to only include the variants that were observed 5 or more times, we retain 80% of observed deletions, which still account for 99% reads with deletions. In substitutions, this filtering retains 28% of all observed substitutions, many of which are sequencing errors. The 'true' TriNEx variants are supported by 81% of all reads containing substitutions.\n",
    "\n",
    "**Table G**: Number of distinct protein variants observed 5 or more times, before sorting the libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eGFP.baseline</th>\n",
       "      <th>GFP8.baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>194</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sd</th>\n",
       "      <td>113</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-dd</th>\n",
       "      <td>191</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sdd</th>\n",
       "      <td>111</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ddd</th>\n",
       "      <td>183</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-sddd</th>\n",
       "      <td>99</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>2118</td>\n",
       "      <td>2220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-ss</th>\n",
       "      <td>820</td>\n",
       "      <td>925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>*</th>\n",
       "      <td>181</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        eGFP.baseline  GFP8.baseline\n",
       "d                 194            186\n",
       "c-sd              113            110\n",
       "c-dd              191            183\n",
       "c-sdd             111            105\n",
       "c-ddd             183            176\n",
       "c-sddd             99            105\n",
       "s                2118           2220\n",
       "c-ss              820            925\n",
       "*                 181            184"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt_prot_count_r, filt_prot_reads_r = get_filtered_protein_composition(results, 5)\n",
    "df = pd.DataFrame(filt_prot_count_r, index=row_index_prot, columns=fractions_all)\n",
    "df[['eGFP.baseline', 'GFP8.baseline']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find where the mutations are located across the gene\n",
    "\n",
    "All three deletion libraries are created starting with insertion of the TransDels transposon, which is an engineered form of the Mu transposon. The transposon duplicates a 5-nt target sequence, which in the case of -3 bp librarires is then *symmetrically* excised to leave behind the 1st and last nt of that sequence. By analysing the frequency of -3 bp insertions we can work backwards and directly observed the positio of insertion of the Mu transposon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_transposon_histogram(all_references, background, baseline='baseline', transposon='d3'):\n",
    "    \"\"\"\n",
    "    Find all mutation of a certain type and count where they are in DNA\n",
    "    :return: dict\n",
    "    \"\"\"\n",
    "    hist = defaultdict(int)\n",
    "    # using a dictionary here allows us to get the positions and counts from the randomly disorganized counts\n",
    "    # in the results dictionary\n",
    "    for prot_mutation in all_references[background][baseline]:\n",
    "        if prot_mutation is None:\n",
    "            continue\n",
    "        elif len(prot_mutation) <= 2:\n",
    "            for dna_mutation, count in all_references[background][baseline][prot_mutation]['dna'].items():\n",
    "                # here we are looking at dna mutations in the tuple format: (position, type, length)\n",
    "                # for deletions, the position in the first bp of tuple is the FIRST position in 1-count (ie. biology standard)\n",
    "                # where the nucleotide is deleted.\n",
    "                # Example: HGVS 74_76del -> (74, d, 3)\n",
    "                if classify_dna(dna_mutation) == transposon:\n",
    "                    hist[int(dna_mutation[0][0])] += count\n",
    "    \n",
    "    # now convert this in an organised list\n",
    "    gene_length = max(hist.keys()) + 1\n",
    "    all_pos = list(range(720))\n",
    "    d3_cnts = [hist[pos] for pos in all_pos] # 'hist' is a default dict: will return 0 for missing values\n",
    "    \n",
    "    \n",
    "    return {'Position': all_pos, 'd3_count': d3_cnts}\n",
    "\n",
    "histogram = find_transposon_histogram(results, 'eGFP', baseline='baseline', transposon='d3')\n",
    "# here, the keys are positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we convert the number of observations per positions to the aggregate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     720.000000\n",
       "mean       55.656944\n",
       "std       148.584870\n",
       "min         0.000000\n",
       "25%         0.000000\n",
       "50%         5.000000\n",
       "75%        44.000000\n",
       "max      2105.000000\n",
       "Name: d3_count, dtype: float64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hist = pd.DataFrame.from_dict(histogram)\n",
    "df_hist['d3_count'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAEICAYAAACzliQjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAfYklEQVR4nO3de5RcZZnv8e/PBJB7AmmYkAsBDIzAGQP2YNQRGVFuIuAsL8lxIDB4og4cRTlLgzqCjKjjCDosFQTJcFEDCDJkAIWIF2TklmCAhICEEEyTkDREIAoigef88b5FdjrV3dVd1VXV2b/PWrV617tvz749+93v3tVbEYGZmZXDa1odgJmZNY+TvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYk46fdD0oWS/qVB05oo6Y+SRuTvv5T04UZMO0/vJ5JmNGp6hek2bB2UnaRJkkLSyPx9SLZZu5H0MUmr8/6/cwOmt1zSOxsR2wDmGZJel7uH7zEREaX9AMuBF4B1wDPAb4CPAq8Z5LTeOcBxfgl8eJCxnwV8v9XrsAHbYCxwMbAS+COwDLgU+OvcfxIQuV/lc1/udyLwci57DlgIHN3qZepneSvLM7KB0zwRuL2Fy9Tnvg9skY+zNzRrnkO0nAG8rlXruVEf1/ThPRGxPbA78FXgM8AljZ5JpWZnG+Qa32+AbYC3AdsDBwK/At7VY/BREbFd/ryhUH5HRGwHjCJtt6sl7TT00dsA7Aq8Flg80BGVOE81UqvPOq38UKW2ABwEvALsn79fCnwpd48BbiBdFawFfk1qIrsij/MCqdb5aTbU6E4Gfg/cRo9aHqmm/xXgbuBZ4Hpgp9zvEKCrWrzAEcBfgJfYuOb7S/KVQ47r88DjwBrgcmDH3K8Sx4wc21PA5/pYT8V1cAjQBZyep7sKOKmPcXckJeNVwBPAl4ARud+XgPvo48qq5zrr0e9ECjVcYNs8bGcN2/5S4NvAjaQrvbuAvQr93wLck7fLPcBb+pjWLODRPJ0HgfcW+o0Avp7X8TLglCr7QGWbnUXh6q3K/nJinsY64DHgQ8DrgT+z4YrnmcLyfQf4SS7/H+CvgG8CfwAeAg4ozGs34FqgO0/744V+ZwFX531oHSl5d+Z+m+z7PdbN3sCf2HC19vP+1m9eJ+fkmF+gSu2adCyckdf3H4D/BF6b+40mHafdud8NwPge+81G67HQ75+AJXm8m4HdC/1erekzgGMC2CrvA78HVgMXAlu3Ku/5DNpDRNxN2oBvq9L79Nyvg1R7+WwaJY4nbdD3RKqJfq0wzttJB+bhvczyBNKOthuwHji/hhh/CnwZuCo2rflWnJg/fw/sCWwHfKvHMH8H7AMcCnxB0uv7m3f2V6RkPo50Uvu2pNG9DHsZableBxwAHAZU7mO8E7guIl6pcb69yldSHyYllkdqHG068EVSklhKSjTkK4UbSdtiZ+A84MY+2qIfJe0vO+bpfV/S2Nzv/wBHk5a9E3jfgBYsk7RtjufISFembwEWRsQSUpPkHXlfGFUY7QOkE/8Y4EXgDuDe/P2avFzkmvR/k07A40j7w2mSivvsMcCVpCuqueR9qZ99n4j4HbBf/joqIt5R4/o9HphJuvp7vJfV8iHScbUX6eTy+Vz+GtJJYHdgIunE8a2+1mPudxzpmP4H0jH+a2BOL/Puqa9j4t9yfFNIx8E44As1TrfhnPSrWwlUayJ4idQGvXtEvBQRv458Ku/DWRHxp4h4oZf+V0TEooj4E/AvwAcqN3rr9CHgvIhYFhF/JNWKpvVoZvpiRLwQEfeRDvhqJ49qXgLOzuvgJlKi3afnQJJ2BY4ETsvrYA3wDWBaHmQM8GRh+GMkPSNpnaRbekzuqdzvGUn/r1A+VdIzeTrTSbXsZ2tcjh9HxN0RsR74AemgBHg38EhEXBER6yNiDqlm/J5qE4mIH0XEyoh4JSKuIp10Dsq9PwB8MyJWRMRa0pXdYL0C7C9p64hYFRH9NZdcFxELIuLPwHXAnyPi8oh4GbiKdCIC+FugIyLOjoi/RMQy0n2WaYVp3R4RN+Vxr6D2faWaWtbvpRGxOPd/qZfpfKuwXs8hbX8i4umIuDYino+Idbnf2wvj9bYePwJ8JSKW5H3iy8AUSbvXsExVjwlJIp34PxkRa3M8X2bjddtUTvrVjSM13/T076Qa4S2SlkmaVcO0Vgyg/+Okm15jaoqyb7uxcQ3pcWAk6Qql4slC9/Okq4FaPJ0Piv7G3Z20PKsqCRv4LrBLZTqkkygAETE311Q/CWzZY1pjImJU/ny9UH5nLhsTEVMj4mc1LgP0vvw91x35+7hqE5F0gqSFhWXcnw3bcDc23cYDlisFHyTV6ldJulHSX/cz2upC9wtVvleWd3dgt8JJ9RlSjbevfeW1ddynqmX99nfc9Bzm8TxdJG0j6buSHpf0HKlpdZSkEf2sx92B/yisg7WA6GW799DbMdFBume1oDDdn+bylnDS70HS35I28u09+0XEuog4PSL2JNVKPiXp0ErvXibZ35XAhEL3RFKN4SlSO+g2hbhGsPGO0t90V5J24uK017PxgT/UVpCaFYoJe4eIqFzu3woc14Y36nquO0jr74meA+Za4MXAqcDO+aS1iJQsILXv9tzGvdlom5OaDF4VETdHxLtIJ8qH8nyh/32hPyuAxwrbaFREbB8RR9U4/kDnX8v6rWWaPdfrytx9OunK800RsQNwcC4X9LkeVwAf6bEeto6I39S4XNU8RTrB7leY5o6RHj5oiXY72FpG0g6Sjia1W34/Ih6oMszRkl6XL9meI908ezn3Xk1qOx+of5S0r6RtgLOBa/Il9O9Ital3S9qC1F65VWG81cCkPhLmHOCTkvaQtB0b7gGs72X4houIVcAtwLl5/b5G0l6SKpfa55Ha06/I5ZK0PRuaWeqSn6s+ZBCj3gTsLel/Sxop6YPAvqQbgj1Vbh5353meRKrpV1wNfFzS+NzG29fV4ULg4Px7jh1JTXKVZdk1N39tSzqR/pGN973xknpeHdXqbuA5SZ+RtLWkEZL2zxWgWgx03x/I+u3LKXm97kS6Mrkql29PSrTP5H5nVkboZz1eCJwhab887I6S3j/AmDaS71ddDHxD0i55uuN63C9pKid9+G9J60hn+c+REtFJvQw7GfgZaUe5A/hORPwy9/sK8Pkqbc79uYL0JMCTpMfaPg6Q26X/GfgeqQb0J9JN5Iof5b9PS7q3ynRn52nfRnpC4c/A/x1AXI1yAqmppvKUxTXkJp2IeAqYmmO7nfQ0xULSQfuxemYqaTxpO21y8u5PRDxNuvl6OqkJ6tOk5/+fqjLsg8C5pP1hNfC/SE+dVFxMegrkPtJN1B/3Md95pMR1P7CAjZPga3I8K0nNDm8n7R8APyc9UfOkpE1i7E+uZLyHdLJ9jFQ7/R7pxmQtBrTvD2T99uOHpErFsvz5Ui7/JrA1aTnuJDWnVPS6HiPiOtJN1ytzs9Ai0j2pen2G1Cx8Z57uz6hyD6xZ1P99SLPhR9I/ki6pz+h34BaSdBvwvYi4vNWxWDn4B0O2WYqI77c6hv7kJr09SbVrs6Zw845ZC+T23SdJvz7e5KEBs6Hi5h0zsxJxTd/MrETavk1/zJgxMWnSpFaHYWY2bCxYsOCpiKj6A7C2T/qTJk1i/vz5rQ7DzGzYkNTrL7/dvGNmViJO+mZmJeKkb2ZWIk76ZmYl4qRvZlYiTvpmZiXipG9mViJO+mZmJeKkb2ZWIk76ZmYNMGnWja0OoSZO+mZmJeKkb2ZWIk76ZmYl4qRvZlYiTvpmZiXipG9mViL9Jn1JEyT9QtISSYslfSKX7yRpnqRH8t/RuVySzpe0VNL9kg4sTGtGHv4RSTOGbrHMzKyaWmr664HTI+L1wFTgFEn7ArOAWyNiMnBr/g5wJDA5f2YCF0A6SQBnAm8CDgLOrJwozMysOfpN+hGxKiLuzd3rgCXAOOBY4LI82GXAcbn7WODySO4ERkkaCxwOzIuItRHxB2AecERDl8bMzPo0oDZ9SZOAA4C7gF0jYhWkEwOwSx5sHLCiMFpXLuutvNp8ZkqaL2l+d3f3QEI0M7M+1Jz0JW0HXAucFhHP9TVolbLoo3zTwoiLIqIzIjo7Oqq+0N3MzAahpqQvaQtSwv9BRPw4F6/OzTbkv2tyeRcwoTD6eGBlH+VmZtYktTy9I+ASYElEnFfoNReoPIEzA7i+UH5CfopnKvBsbv65GThM0uh8A/ewXGZmZk0ysoZh3gocDzwgaWEu+yzwVeBqSScDvwfen/vdBBwFLAWeB04CiIi1kv4VuCcPd3ZErG3IUpiZWU36TfoRcTvV2+MBDq0yfACn9DKt2cDsgQRoZmaN41/kmpmViJO+mVmJOOmbmZWIk76ZWYk46ZuZlYiTvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYk46ZuZlYiTvplZiTjpm5mViJO+mVmJOOmbmZVILW/Omi1pjaRFhbKrJC3Mn+WVl6tImiTphUK/CwvjvFHSA5KWSjo/v5HLzMyaqJY3Z10KfAu4vFIQER+sdEs6F3i2MPyjETGlynQuAGYCd5LernUE8JOBh2xmZoPVb00/Im4Dqr7WMNfWPwDM6Wsa+cXpO0TEHfnNWpcDxw08XDMzq0e9bfpvA1ZHxCOFsj0k/VbSryS9LZeNA7oKw3TlMjMza6Jamnf6Mp2Na/mrgIkR8bSkNwL/JWk/qr9jN3qbqKSZpKYgJk6cWGeIZmZWMeiavqSRwD8AV1XKIuLFiHg6dy8AHgX2JtXsxxdGHw+s7G3aEXFRRHRGRGdHR8dgQzQzsx7qad55J/BQRLzabCOpQ9KI3L0nMBlYFhGrgHWSpub7ACcA19cxbzMzG4RaHtmcA9wB7COpS9LJudc0Nr2BezBwv6T7gGuAj0ZE5Sbwx4DvAUtJVwB+csfMrMn6bdOPiOm9lJ9Ypexa4Npehp8P7D/A+MzMrIH8i1wzsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrkVrenDVb0hpJiwplZ0l6QtLC/Dmq0O8MSUslPSzp8EL5EblsqaRZjV8UMzPrTy01/UuBI6qUfyMipuTPTQCS9iW9RnG/PM53JI3I7839NnAksC8wPQ9rZmZNVMvrEm+TNKnG6R0LXBkRLwKPSVoKHJT7LY2IZQCSrszDPjjgiM3MbNDqadM/VdL9uflndC4bB6woDNOVy3orr0rSTEnzJc3v7u6uI0QzMysabNK/ANgLmAKsAs7N5aoybPRRXlVEXBQRnRHR2dHRMcgQzcysp36bd6qJiNWVbkkXAzfkr13AhMKg44GVubu3cjMza5JB1fQljS18fS9QebJnLjBN0laS9gAmA3cD9wCTJe0haUvSzd65gw/bzMwGo9+avqQ5wCHAGEldwJnAIZKmkJpolgMfAYiIxZKuJt2gXQ+cEhEv5+mcCtwMjABmR8Tihi+NmZn1qZand6ZXKb6kj+HPAc6pUn4TcNOAojMzs4byL3LNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSsRJ38ysRPpN+pJmS1ojaVGh7N8lPSTpfknXSRqVyydJekHSwvy5sDDOGyU9IGmppPMlVXtZupmZDaFaavqXAkf0KJsH7B8RfwP8Djij0O/RiJiSPx8tlF8AzCS9N3dylWmamdkQ6zfpR8RtwNoeZbdExPr89U5gfF/TyC9S3yEi7oiIAC4HjhtcyGZmNliNaNP/J+Anhe97SPqtpF9JelsuGwd0FYbpymVVSZopab6k+d3d3Q0I0czMoM6kL+lzwHrgB7loFTAxIg4APgX8UNIOQLX2++htuhFxUUR0RkRnR0dHPSGamVnByMGOKGkGcDRwaG6yISJeBF7M3QskPQrsTarZF5uAxgMrBztvMzMbnEHV9CUdAXwGOCYini+Ud0gakbv3JN2wXRYRq4B1kqbmp3ZOAK6vO3ozMxuQfmv6kuYAhwBjJHUBZ5Ke1tkKmJefvLwzP6lzMHC2pPXAy8BHI6JyE/hjpCeBtibdAyjeBzAzsyboN+lHxPQqxZf0Muy1wLW99JsP7D+g6MzMrKH8i1wzsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEqkp6UuaLWmNpEWFsp0kzZP0SP47OpdL0vmSlkq6X9KBhXFm5OEfya9bNDOzJqq1pn8pcESPslnArRExGbg1fwc4kvSaxMnATOACSCcJ0lu33gQcBJxZOVGYmVlz1JT0I+I2YG2P4mOBy3L3ZcBxhfLLI7kTGCVpLHA4MC8i1kbEH4B5bHoiMTOzIVRPm/6u+YXn5L+75PJxwIrCcF25rLfyTUiaKWm+pPnd3d11hGhmZkVDcSNXVcqij/JNCyMuiojOiOjs6OhoaHBmZmVWT9JfnZttyH/X5PIuYEJhuPHAyj7Kzcw2K5Nm3djqEHpVT9KfC1SewJkBXF8oPyE/xTMVeDY3/9wMHCZpdL6Be1guMzOzJhlZy0CS5gCHAGMkdZGewvkqcLWkk4HfA+/Pg98EHAUsBZ4HTgKIiLWS/hW4Jw93dkT0vDlsZmZDqKakHxHTe+l1aJVhAzill+nMBmbXHJ2ZmTWUf5FrZlYiTvpmZiXipG9mViJO+mZmJeKkb2ZWIk76ZmYl4qRvZlYiTvpmZiXipG9mViJO+mZmJeKkb2ZWIk76Zm2knf8lr20enPTNzErESd/MrESc9M3MSsRJ38ysRAad9CXtI2lh4fOcpNMknSXpiUL5UYVxzpC0VNLDkg5vzCKYmVmtanpzVjUR8TAwBUDSCOAJ4DrS6xG/ERFfLw4vaV9gGrAfsBvwM0l7R8TLg43BzMwGplHNO4cCj0bE430McyxwZUS8GBGPkd6he1CD5m9mZjVoVNKfBswpfD9V0v2SZksancvGASsKw3Tlsk1ImilpvqT53d3dDQrRzMzqTvqStgSOAX6Uiy4A9iI1/awCzq0MWmX0qDbNiLgoIjojorOjo6PeEM3MLGtETf9I4N6IWA0QEasj4uWIeAW4mA1NOF3AhMJ444GVDZi/mZnVqBFJfzqFph1JYwv93gssyt1zgWmStpK0BzAZuLsB8zczsxoN+ukdAEnbAO8CPlIo/pqkKaSmm+WVfhGxWNLVwIPAeuAUP7ljZtZcdSX9iHge2LlH2fF9DH8OcE498zQzs8HzL3LNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSqQR78hdLukBSQslzc9lO0maJ+mR/Hd0Lpek8yUtzS9OP7De+ZuZWe0aVdP/+4iYEhGd+fss4NaImAzcmr9Dep/u5PyZSXqJupmZNclQNe8cC1yWuy8DjiuUXx7JncCoHu/UNTOzIdSIpB/ALZIWSJqZy3aNiFUA+e8uuXwcsKIwblcuMzOzJqjrHbnZWyNipaRdgHmSHupjWFUpi00GSiePmQATJ05sQIhmZgYNqOlHxMr8dw1wHXAQsLrSbJP/rsmDdwETCqOPB1ZWmeZFEdEZEZ0dHR31hmhmZlldSV/StpK2r3QDhwGLgLnAjDzYDOD63D0XOCE/xTMVeLbSDGRmZkOv3uadXYHrJFWm9cOI+Kmke4CrJZ0M/B54fx7+JuAoYCnwPHBSnfM3M7MBqCvpR8Qy4A1Vyp8GDq1SHsAp9czTzMwGz7/INTMrESd9M7MScdK3YWfSrBtbHYLZsOWkb2ZWIk76ZmYl4qRvZlYiTvpmZiXipG9mViJO+mZtzk8rWSM56ZuZlYiTvplZiTjpmw0TbuaxRnDSNzMrESd9G1Zc220f3hbDk5O+mVmJOOmbNZFrx9Zqg076kiZI+oWkJZIWS/pELj9L0hOSFubPUYVxzpC0VNLDkg5vxAKYmQ2lze1EXc+bs9YDp0fEvfk9uQskzcv9vhERXy8OLGlfYBqwH7Ab8DNJe0fEy3XEYGZmAzDomn5ErIqIe3P3OmAJMK6PUY4FroyIFyPiMdJ7cg8a7PzNWmFzq/UNB17njdWQNn1Jk4ADgLty0amS7pc0W9LoXDYOWFEYrYteThKSZkqaL2l+d3d3I0I0GxQnHIPNaz+oO+lL2g64FjgtIp4DLgD2AqYAq4BzK4NWGT2qTTMiLoqIzojo7OjoqDdEG6TNaUc3s6SupC9pC1LC/0FE/BggIlZHxMsR8QpwMRuacLqACYXRxwMr65m/mZkNTD1P7wi4BFgSEecVyscWBnsvsCh3zwWmSdpK0h7AZODuwc7fzKrzFZr1pZ6nd94KHA88IGlhLvssMF3SFFLTzXLgIwARsVjS1cCDpCd/TvGTO2aNUUn0y7/67hZHYu1u0Ek/Im6nejv9TX2Mcw5wzmDnaWZm9fEvcm2z5+YOsw2c9K1tDadkPZxibbZJs24cdutnuMU7EE76ZlaTzTkRlomT/iD5ALC+eP9IvB7aj5N+G2vFAeODtHmK69rrvTGGw3psdYxO+rZZa/UBZtZunPRbzEmp8fpap2Ve32Va9jIt60A56VvdfICZDR9O+mZWs+Fwgh8OMbaSk/5maKh2eh9Mw4+3WXtq5XZx0jcbxpzUh4922VZO+tYQve3Q/e3o7XIgtIN2XhftHJsNjJP+MDIcDrx6Y6yMPxyWdXPX6G3Qrtu0XeMaKk76NmwN9GBt5MHdqJObWbM56ZeYE09tNuf1VG3ZNuflNSd9GwKtrIEPVqPuPbT6h2HtsC4bpZ2XpZ1j60/Tk76kIyQ9LGmppFnNnn+76O3/rrRyZ2pEYmvXcQcznVadvIbjvyKG9m9uq6zXZs23Xe9PNTXpSxoBfBs4EtiX9GrFfZsZQzXttlGKhiK2ntPsb8cdTBNAMxNALQdyrcvQ6Jp6O7f997UftOqf/Q3XE15/2mmZml3TPwhYGhHLIuIvwJXAsU2OoSaDPfgH0q/WYat1V5tWX8P19rfWWAczXDOaOoYqqdZac6ynf63D9DZOo2qSjdgWxVja6Yb5cJhns68IFBFNmRGApPcBR0TEh/P344E3RcSpPYabCczMX/cBHh7kLMcATw1y3GYaLnHC8InVcTbWcIkThk+sQxnn7hHRUa3HoF+MPkjVXqS+yVknIi4CLqp7ZtL8iOisdzpDbbjECcMnVsfZWMMlThg+sbYqzmY373QBEwrfxwMrmxyDmVlpNTvp3wNMlrSHpC2BacDcJsdgZlZaTW3eiYj1kk4FbgZGALMjYvEQzrLuJqImGS5xwvCJ1XE21nCJE4ZPrC2Js6k3cs3MrLX8i1wzsxJx0jczK5HNNum30797kDRb0hpJiwplO0maJ+mR/Hd0Lpek83Pc90s6sIlxTpD0C0lLJC2W9Il2jFXSayXdLem+HOcXc/keku7KcV6VHxZA0lb5+9Lcf1Iz4izEO0LSbyXd0OZxLpf0gKSFkubnsrba9nneoyRdI+mhvK++ud3ilLRPXo+Vz3OSTmuLOCNis/uQbhI/CuwJbAncB+zbwngOBg4EFhXKvgbMyt2zgH/L3UcBPyH9pmEqcFcT4xwLHJi7twd+R/p3GW0Va57fdrl7C+CuPP+rgWm5/ELgY7n7n4ELc/c04Komb/9PAT8Ebsjf2zXO5cCYHmVtte3zvC8DPpy7twRGtWOchXhHAE8Cu7dDnE1d+Cau5DcDNxe+nwGc0eKYJvVI+g8DY3P3WODh3P1dYHq14VoQ8/XAu9o5VmAb4F7gTaRfN47suQ+QnhZ7c+4emYdTk+IbD9wKvAO4IR/UbRdnnme1pN9W2x7YAXis53pptzh7xHYY8D/tEufm2rwzDlhR+N6Vy9rJrhGxCiD/3SWXt0XsuWnhAFItuu1izU0mC4E1wDzSld0zEbG+Siyvxpn7Pwvs3Iw4gW8CnwZeyd93btM4If06/hZJC5T+FQq037bfE+gG/jM3mX1P0rZtGGfRNGBO7m55nJtr0q/p3z20qZbHLmk74FrgtIh4rq9Bq5Q1JdaIeDkippBq0gcBr+8jlpbEKeloYE1ELCgW9xFLq7f9WyPiQNJ/wT1F0sF9DNuqWEeSmkoviIgDgD+Rmkl609J1mu/XHAP8qL9Bq5QNSZyba9IfDv/uYbWksQD575pc3tLYJW1BSvg/iIgft3OsABHxDPBLUjvoKEmVHxwWY3k1ztx/R2BtE8J7K3CMpOWk/yj7DlLNv93iBCAiVua/a4DrSCfTdtv2XUBXRNyVv19DOgm0W5wVRwL3RsTq/L3lcW6uSX84/LuHucCM3D2D1H5eKT8h382fCjxbuRwcapIEXAIsiYjz2jVWSR2SRuXurYF3AkuAXwDv6yXOSvzvA34eueF0KEXEGRExPiImkfbBn0fEh9otTgBJ20ravtJNaodeRJtt+4h4ElghaZ9cdCjwYLvFWTCdDU07lXhaG2czb2g0+ebJUaSnTx4FPtfiWOYAq4CXSGf0k0lttbcCj+S/O+VhRXrRzKPAA0BnE+P8O9Il5f3Awvw5qt1iBf4G+G2OcxHwhVy+J3A3sJR0Ob1VLn9t/r4099+zBfvAIWx4eqft4swx3Zc/iyvHTLtt+zzvKcD8vP3/CxjdpnFuAzwN7Fgoa3mc/jcMZmYlsrk275iZWRVO+mZmJeKkb2ZWIk76ZmYl4qRvZlYiTvpmZiXipG9mViL/H3DDHIwhfWzLAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.bar(df_hist['Position'], df_hist['d3_count'], width=1)\n",
    "plt.title('Distribution in eGFP, no adjustment for baseline')\n",
    "plt.show()\n",
    "\n",
    "# make this a prettier histogram with a split axis that lets you see all positions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many deletions per position are detected?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insertion_frequencies(results, background):\n",
    "\n",
    "    ins_freq = {deletion: defaultdict(int) for deletion in ['d3','d6','d9']}  # the starting data is aggregated in the 'baseline' fraction\n",
    "\n",
    "    for prot_mutation, data in results[background]['baseline'].items():  # this will go through all protein mutations\n",
    "        for dna_mutation, c in data['dna'].items():\n",
    "            # sometimes a mutations will have a sequencing error / silent substitution elsewhere, which biases the count\n",
    "            dna_mut_type = classify_dna(dna_mutation)\n",
    "            if dna_mut_type in ['d3', 'd6', 'd9']:\n",
    "                prot_mut_type = classify_protein(prot_mutation)\n",
    "                if prot_mut_type in translate_variant_type(dna_mut_type):\n",
    "                    ins_freq[dna_mut_type][int(dna_mutation[0][0])] += 1\n",
    "    \n",
    "    position = list(range(720))\n",
    "    d3_cnts = [ins_freq['d3'][pos] for pos in position]\n",
    "    d6_cnts = [ins_freq['d6'][pos] for pos in position]\n",
    "    d9_cnts = [ins_freq['d9'][pos] for pos in position]\n",
    "    \n",
    "    return {'Position': position, 'd3': d3_cnts, 'd6': d6_cnts, 'd9': d9_cnts}\n",
    "    \n",
    "ins_freq = insertion_frequencies(results, 'eGFP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlMAAADSCAYAAABuMkW8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAW7UlEQVR4nO3de7BlZXnn8e8vdItEUNA+DsjFI0qIYgXBFmE0KUZNRLwwlfECM15Lq6ssSaTC1AjODF4mmcKqjBrB0WBE0BBvQJzmYpCgjpdM0KYDCHSQRhq70ygtCEi8Nj7zx14HNpt9O71299nnnO+natdZ633f9a5nPefsfZ5ae+29UlVIkiRpx/zGQgcgSZK0mFlMSZIktWAxJUmS1ILFlCRJUgsWU5IkSS1YTEmSJLVgMSVpWUhybJItXes3Jjl2AUOStERYTEnLVJJDknwmybYk9yW5JclZSQ5o+o9N8usk93c9Lmn63p3kV03bPUn+IckxC3tE81NVh1XVV9vM0eThrycU0o7sv5I8baH2L6nDYkpahpp/wFcDW4EjquqxwPOAW4Hndw3dWlV7dj1e3tX32araE5gBvgFcnCS76BAkaWpYTElLVJInJbmoOfN0W5I/7up+N/DNqvqTqtoCUFV3VtUHq+oz89lPVf0KOB/YF3jCGHF9Ncn/SPLNJD9J8qUkq7r6X9G8BXdPM/bpQ+b6iySbmzNr1yT53a6+PZKcl+THSW4CntOz7aYkL2qWz0vyp119vW8JviPJvzTx3pzkhUmOA94JvKY5Q3dd1/H9aXO27v4klyR5QpILmji/nWS2a+7fTnJlkrubuV/d1Xdekg8nuazZ99VJntr0fa0Zdl2zn9eMyr2kncNiSlqCkvwGcAlwHbA/8ELglCQvboa8CLhoQvvaHXgjsKWqfjTmZv8ReBPwROBRwH9u5vot4NPAKXTOeF0OXJLkUQPm+TbwLODxwN8An0/y6KbvXcBTm8eLgTfM68AaSQ4FTgaeU1V7NXNtqqq/A/4nzRm6qjq8a7MTgdfRyf1Tgf8HfKKJc0MTG0keA1zZxP5E4CTgfyc5rGuuk4D3APsAG4E/A6iq32v6D2/2/9kdOT5J7VlMSUvTc4CZqnpvVf2yqr4HfIzOP3mAVcAP5gYnObk5E3R/ko91zfOkpn3u8equvlcnuQfYDDwb+PfziO8TVfXdqvoZ8Dk6BRHAa4DLqurK5ozXnwN7AP+23yRV9ddVdVdVba+q/wXsDhw6Fx/wZ1V1d1VtBj40j/i6PdDM+4wkK6tqU1XdOsbx3VpV9wJfBG6tqr+vqu3A54EjmnEvo1OYfaI5hvV0itxXds11cVV9q9n2Ah7KlaQpsWKhA5C0UzyZphDqatsN+HqzfBew31xHVZ0NnN281XVA1zZbq6p7vdvnquq1OxjfD7qWfwrs2Sw/Cbi9K65fJ9lM5wzPIyQ5FXhLs10Bj6VTKM7Ntblr+O3sgKramOQUOm+NHpbkCuBPqmrrkM1+2LX8sz7rc8f7ZOC5Pb+nFcCnutYH5UrSlPDMlLQ0bQZuq6q9ux57VdXxTf9VwB8uYHyDbKVTYADQXNB+IPAvvQOb66PeQecM1D5VtTdwLzB3EfwdzbZzDhqy338FfrNrfd/uzqr6m6p6fhNbAe+b6xp9SENtBv5vz+9pz6p6a8t5Je1CFlPS0vQt4L7mwuk9kuyW5JlJ5i7Cfjfwu0nen2R/gOYi8IEXe48ryWzzkf3ZHdj8c8BLmwu8VwKnAr8A/qHP2L2A7cA2YEWSM+icmeqe6/Qk+zRf9/BHQ/Z7LXB8kscn2ZfONVtzx3Nokhc014b9nM6ZpQea7h8Cs801ajviUuC3krwuycrm8ZxhF933+CFw8A7uW9KEWExJS1BVPQC8nM71NbcBPwL+Cnhc0/9d4Gg6b+ldl+QnwDfpnBn67y13fyCdt9QecTZpjLhvBl4LnNXE/HLg5VX1yz7Dr6BzPdJ3m/39nIe/rfeepv024Es8/K2zXp+ic7H+pmZs98XcuwNnNvH8gM6F4u9s+j7f/LwryfpxjrFbVf0E+AM617JtbeZ/X7PPcbwbOL/P9WySdqFUtT1LLUkPSfLfgG1V9ZcLHcswSb4PvLaqvjZysCQNYTEladlJMgN8Hzi0qr6/0PFIWtx8m0/SstJcN3YLcJaFlKRJ8MyUJElSC56ZkiRJasFiSpIkqYUF+wb0VatW1ezs7ELtXpIkaWzXXHPNj6pqpl/fghVTs7OzrFu3bqF2L0mSNLYkA29J5dt8kiRJLVhMSZIktTCymEry6CTfSnJdkhuTvKfPmN2TfDbJxiRX7+A9uSRJkhadcc5M/QJ4QVUdTuc+X8clObpnzJuBH1fV04AP8NAd1SVJkpa0kcVUddzfrK5sHr3f9HkCcH6zfCHwwiSZWJSSJElTaqxrppLsluRa4E7gyqq6umfI/jR3a6+q7cC9wBP6zLMmybok67Zt29Yu8ik3e9plA9t7+0atj7OvuW36zT9OTOO0d6/322ZY3KO2Hcc4uetuG5SLcfM1aB/zGd+9n3HmHqe/d/5B+xg0blCM48bZb3/z+Zuei2XY72bQmN59j/r9DdvPOG399jmf7YZtO+w4h+1jR/6mR/3Oxn3uztc4z8VR24/zexnnb3LQfsd9Dg1rn+9r8I7mYlT/OHmez3O83/hhryuDYhrWNuo1fLEYq5iqqgeq6lnAAcBRSZ7ZM6TfWahH3Kemqs6pqtVVtXpmpu9XNUiSJC0q8/o0X1XdA3wVOK6nawtwIECSFcDjgLsnEJ8kSdJUG+fTfDNJ9m6W9wBeBPxzz7C1wBua5VcCXy7voCxJkpaBcb4BfT/g/CS70Sm+PldVlyZ5L7CuqtYCHwc+lWQjnTNSJ+60iCVJkqbIyGKqqq4HjujTfkbX8s+BV002NEmSpOnnN6BLkiS1YDElSZLUgsWUJElSCxZTkiRJLVhMSZIktWAxJUmS1ILFlCRJUgsWU5IkSS1YTEmSJLVgMSVJktSCxZQkSVILFlOSJEktWExJkiS1YDElSZLUgsWUJElSCxZTkiRJLVhMSZIktTCymEpyYJKvJNmQ5MYkb+8z5tgk9ya5tnmcsXPClSRJmi4rxhizHTi1qtYn2Qu4JsmVVXVTz7ivV9XLJh+iJEnS9Bp5Zqqq7qiq9c3yT4ANwP47OzBJkqTFYF7XTCWZBY4Aru7TfUyS65J8MclhA7Zfk2RdknXbtm2bd7CSJEnTZuxiKsmewEXAKVV1X0/3euDJVXU4cBbwhX5zVNU5VbW6qlbPzMzsaMySJElTY6xiKslKOoXUBVV1cW9/Vd1XVfc3y5cDK5OsmmikkiRJU2icT/MF+DiwoareP2DMvs04khzVzHvXJAOVJEmaRuN8mu95wOuA7yS5tml7J3AQQFV9FHgl8NYk24GfASdWVe2EeCVJkqbKyGKqqr4BZMSYs4GzJxWUJEnSYuE3oEuSJLVgMSVJktSCxZQkSVILFlOSJEktWExJkiS1YDElSZLUgsWUJElSCxZTkiRJLVhMSZIktWAxJUmS1ILFlCRJUgsWU5IkSS1YTEmSJLVgMSVJktSCxZQkSVILFlOSJEktWExJkiS1MLKYSnJgkq8k2ZDkxiRv7zMmST6UZGOS65McuXPClSRJmi4rxhizHTi1qtYn2Qu4JsmVVXVT15iXAIc0j+cCH2l+SpIkLWkjz0xV1R1Vtb5Z/gmwAdi/Z9gJwCer4x+BvZPsN/FoJUmSpsw4Z6YelGQWOAK4uqdrf2Bz1/qWpu2Onu3XAGsADjrooPlFugBmT7uMTWe+lNnTLgNg05kvfVh777gdnX8+++43btj84+5/3HkGbTMoH/1yNWh//bYZdQzdY0b9Dnrn6ndcvWOGHddcf7/2YcfQb1/dbf3m7Dd+1LH0Lg+La5xc9Pt7nU+cg36X8/kdDxo/LH/9+kfFM87YYb+D+cTeb9/D1nvnGfe1YVR8/XI4aPve/mHPvVF/R8PmHTf2Qe3jPp/7bT/uc32cePrFNOz1s7t9WCz9th90HP32Pei50m/bUX3D9tt7XL1jxtnXjv6f3VXGvgA9yZ7ARcApVXVfb3efTeoRDVXnVNXqqlo9MzMzv0glSZKm0FjFVJKVdAqpC6rq4j5DtgAHdq0fAGxtH54kSdJ0G+fTfAE+DmyoqvcPGLYWeH3zqb6jgXur6o4BYyVJkpaMca6Zeh7wOuA7Sa5t2t4JHARQVR8FLgeOBzYCPwXeNPlQJUmSps/IYqqqvkH/a6K6xxTwtkkFJUmStFj4DeiSJEktWExJkiS1YDElSZLUgsWUJElSCxZTkiRJLVhMSZIktWAxJUmS1ILFlCRJUgsWU5IkSS1YTEmSJLVgMSVJktSCxZQkSVILFlOSJEktWExJkiS1YDElSZLUgsWUJElSCxZTkiRJLYwsppKcm+TOJDcM6D82yb1Jrm0eZ0w+TEmSpOm0Yowx5wFnA58cMubrVfWyiUQkSZK0iIw8M1VVXwPu3gWxSJIkLTqTumbqmCTXJfliksMGDUqyJsm6JOu2bds2oV1LkiQtnEkUU+uBJ1fV4cBZwBcGDayqc6pqdVWtnpmZmcCuJUmSFlbrYqqq7quq+5vly4GVSVa1jkySJGkRaF1MJdk3SZrlo5o572o7ryRJ0mIw8tN8ST4NHAusSrIFeBewEqCqPgq8Enhrku3Az4ATq6p2WsSSJElTZGQxVVUnjeg/m85XJ0iSJC07fgO6JElSCxZTkiRJLVhMSZIktWAxJUmS1ILFlCRJUgsWU5IkSS1YTEmSJLVgMSVJktSCxZQkSVILFlOSJEktWExJkiS1YDElSZLUgsWUJElSCxZTkiRJLVhMSZIktWAxJUmS1ILFlCRJUgsji6kk5ya5M8kNA/qT5ENJNia5PsmRkw9TkiRpOo1zZuo84Lgh/S8BDmkea4CPtA9LkiRpcRhZTFXV14C7hww5AfhkdfwjsHeS/SYVoCRJ0jRLVY0elMwCl1bVM/v0XQqcWVXfaNavAt5RVev6jF1D5+wVBx100LNvv/32VsHviNnTLmPTmS+d19jZ0y57sK13vbt92Laj5hk0b7/+7n0N22bYPKP219bOnn/UPnZk/5OOeVfkQA8ZN9+T/ruZlF353IT5v3b0zjFq+/m8pk3CrsrffONZ6LmnJR+TMu7/750hyTVVtbpf3yQuQE+ftr4VWlWdU1Wrq2r1zMzMBHYtSZK0sCZRTG0BDuxaPwDYOoF5JUmSpt4kiqm1wOubT/UdDdxbVXdMYF5JkqSpt2LUgCSfBo4FViXZArwLWAlQVR8FLgeOBzYCPwXetLOClSRJmjYji6mqOmlEfwFvm1hEkiRJi4jfgC5JktSCxZQkSVILFlOSJEktWExJkiS1YDElSZLUgsWUJElSCxZTkiRJLVhMSZIktWAxJUmS1ILFlCRJUgsWU5IkSS1YTEmSJLVgMSVJktSCxZQkSVILFlOSJEktWExJkiS1MFYxleS4JDcn2ZjktD79b0yyLcm1zeMtkw9VkiRp+qwYNSDJbsCHgd8HtgDfTrK2qm7qGfrZqjp5J8QoSZI0tcY5M3UUsLGqvldVvwQ+A5ywc8OSJElaHMYppvYHNnetb2naev2HJNcnuTDJgROJTpIkacqNU0ylT1v1rF8CzFbV7wB/D5zfd6JkTZJ1SdZt27ZtfpFKkiRNoXGKqS1A95mmA4Ct3QOq6q6q+kWz+jHg2f0mqqpzqmp1Va2emZnZkXglSZKmyjjF1LeBQ5I8JcmjgBOBtd0DkuzXtfoKYMPkQpQkSZpeIz/NV1Xbk5wMXAHsBpxbVTcmeS+wrqrWAn+c5BXAduBu4I07MWZJkqSpMbKYAqiqy4HLe9rO6Fo+HTh9sqFJkiRNP78BXZIkqQWLKUmSpBYspiRJklqwmJIkSWrBYkqSJKkFiylJkqQWLKYkSZJasJiSJElqwWJKkiSpBYspSZKkFiymJEmSWrCYkiRJasFiSpIkqQWLKUmSpBYspiRJklqwmJIkSWrBYkqSJKmFsYqpJMcluTnJxiSn9enfPclnm/6rk8xOOlBJkqRpNLKYSrIb8GHgJcAzgJOSPKNn2JuBH1fV04APAO+bdKCSJEnTaJwzU0cBG6vqe1X1S+AzwAk9Y04Azm+WLwRemCSTC1OSJGk6jVNM7Q9s7lrf0rT1HVNV24F7gSdMIkBJkqRplqoaPiB5FfDiqnpLs/464Kiq+qOuMTc2Y7Y067c2Y+7qmWsNsKZZPRS4eVIHMsQq4Ee7YD+LjXkZzNz0Z14GMzf9mZfBzE1/05yXJ1fVTL+OFWNsvAU4sGv9AGDrgDFbkqwAHgfc3TtRVZ0DnDNOxJOSZF1Vrd6V+1wMzMtg5qY/8zKYuenPvAxmbvpbrHkZ522+bwOHJHlKkkcBJwJre8asBd7QLL8S+HKNOuUlSZK0BIw8M1VV25OcDFwB7AacW1U3JnkvsK6q1gIfBz6VZCOdM1In7sygJUmSpsU4b/NRVZcDl/e0ndG1/HPgVZMNbWJ26duKi4h5Gczc9GdeBjM3/ZmXwcxNf4syLyMvQJckSdJg3k5GkiSphSVbTI26Bc5Sl+TcJHcmuaGr7fFJrkxyS/Nzn6Y9ST7U5Or6JEcuXOQ7V5IDk3wlyYYkNyZ5e9NubpJHJ/lWkuua3LynaX9Kc5uoW5rbRj2qaV9Wt5FKsluSf0pyabNuXoAkm5J8J8m1SdY1bT6fkr2TXJjkn5vXm2OWe16SHNr8ncw97ktyylLIy5IspjLeLXCWuvOA43raTgOuqqpDgKuadejk6ZDmsQb4yC6KcSFsB06tqqcDRwNva/42zA38AnhBVR0OPAs4LsnRdG4P9YEmNz+mc/soWH63kXo7sKFr3bw85N9V1bO6PtLu8wn+Avi7qvpt4HA6fzvLOi9VdXPzd/Is4NnAT4G/ZSnkpaqW3AM4Briia/104PSFjmsB8jAL3NC1fjOwX7O8H3Bzs/yXwEn9xi31B/B/gN83N4/Iy28C64Hn0vkCvRVN+4PPLTqf8D2mWV7RjMtCx76T8nEAnRf5FwCXAjEvD+ZmE7Cqp21ZP5+AxwK39f7el3teenLxB8A3l0peluSZKca7Bc5y9G+q6g6A5ucTm/Zlma/m7ZcjgKsxN8CDb2VdC9wJXAncCtxTndtEwcOPfzndRuqDwH8Bft2sPwHzMqeALyW5Jp27XIDPp4OBbcAnmreG/yrJYzAv3U4EPt0sL/q8LNViqt9Nlv3Y4mDLLl9J9gQuAk6pqvuGDe3TtmRzU1UPVOcU/AF0bnL+9H7Dmp/LIjdJXgbcWVXXdDf3Gbqs8tLleVV1JJ23ZN6W5PeGjF0uuVkBHAl8pKqOAP6Vh9666me55AWA5vrCVwCfHzW0T9tU5mWpFlPj3AJnOfphkv0Amp93Nu3LKl9JVtIppC6oqoubZnPTparuAb5K57qyvdO5TRQ8/PgfzE2G3EZqCXge8Iokm4DP0Hmr74OYFwCqamvz8046178chc+nLcCWqrq6Wb+QTnG13PMy5yXA+qr6YbO+6POyVIupcW6Bsxx13/bnDXSuF5prf33zyYmjgXvnTrkuNUlC5xv7N1TV+7u6zE0yk2TvZnkP4EV0Lpr9Cp3bRMEjc7PkbyNVVadX1QFVNUvnteTLVfWfWOZ5AUjymCR7zS3TuQ7mBpb586mqfgBsTnJo0/RC4CaWeV66nMRDb/HBUsjLQl+0tbMewPHAd+lc8/FfFzqeBTj+TwN3AL+iU92/mc51G1cBtzQ/H9+MDZ1PP94KfAdYvdDx78S8PJ/OaeLrgWubx/HmpgB+B/inJjc3AGc07QcD3wI20jktv3vT/uhmfWPTf/BCH8MuyNGxwKXm5cF8HAxc1zxunHut9flU0PlE7Lrm+fQFYB/z8uCHW+4CHtfVtujz4jegS5IktbBU3+aTJEnaJSymJEmSWrCYkiRJasFiSpIkqQWLKUmSpBYspiRJklqwmJIkSWrBYkqSJKmF/w+7X53DN9/n3wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_ins_freq = pd.DataFrame.from_dict(ins_freq)\n",
    "df_ins_freq['dels'] = df_ins_freq['d3'] + df_ins_freq['d6'] + df_ins_freq['d9']\n",
    "df_ins_freq\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 3))\n",
    "ax.bar(df_ins_freq['Position'], df_ins_freq['dels'], width=1)\n",
    "ax.set_title('eGFP, no adjustment')\n",
    "plt.show()\n",
    "# df_ins_freq.to_csv('Fig3BC_all_ref.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2    37.777778\n",
       "1    25.833333\n",
       "3    25.138889\n",
       "0    11.250000\n",
       "Name: dels, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dels_agg = df_ins_freq['dels'].value_counts()\n",
    "dels_agg/720*100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see that 25% of positions have all three positions and 11% are entirely unobserved. This must be adjusted by taking into account the baseline."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look for the most common variants in the High dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a Counter dictionary just with dna_hgvs mutations and counts\n",
    "# Here, associate just the HGVS DNA variant description and the count\n",
    "\n",
    "gfp_dna = {'eGFP': {}, 'GFP8': {}}\n",
    "for bg in results.keys():\n",
    "    for fraction in results[bg].keys():\n",
    "        gfp_dna[bg][fraction] = {}\n",
    "        for pm, pdata in results[bg][fraction].items():  # pm - protein mutation\n",
    "            for dna, reads in pdata['dna_hgvs'].items():\n",
    "                gfp_dna[bg][fraction][dna] = reads\n",
    "        del gfp_dna[bg][fraction][None]\n",
    "        del gfp_dna[bg][fraction][str(bg) + ':c.[]']\n",
    "\n",
    "# once we're done extraction all information, convert to Counter\n",
    "\n",
    "gfp_cnt = {'eGFP': {}, 'GFP8': {}}\n",
    "for bg in gfp_cnt.keys():\n",
    "    for fraction in results[bg].keys():\n",
    "        gfp_cnt[bg][fraction] = Counter(gfp_dna[bg][fraction])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working with a simpler `Counter` for DNA/read statistics is useful for converting to a dataframe, but the linkage between the DNA variant and the effect when translated to protein is lost. Make up for this by making a separate \"translation\" dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper dictionary: DNA_hgvs to protein\n",
    "\n",
    "gfp_translate = {'eGFP': {}, 'GFP8': {}}\n",
    "\n",
    "for bg in results.keys():\n",
    "    for fraction in results[bg].keys():\n",
    "        gfp_dna[bg][fraction] = {}\n",
    "        for pm, pdata in results[bg][fraction].items():  # pm - protein mutation\n",
    "            for dna in pdata['dna_hgvs'].keys():\n",
    "                try:\n",
    "                    gfp_translate[bg][dna]\n",
    "                except KeyError:\n",
    "                    gfp_translate[bg][dna] = pm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_hgvs</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Baseline</th>\n",
       "      <th>High</th>\n",
       "      <th>Medium</th>\n",
       "      <th>Low</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>eGFP:c.674_676del</td>\n",
       "      <td>((226, 'd'),)</td>\n",
       "      <td>2105</td>\n",
       "      <td>9468</td>\n",
       "      <td>340</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>eGFP:c.458_463del</td>\n",
       "      <td>((153, 's', 'N'), (154, 'd'), (155, 'd'))</td>\n",
       "      <td>1247</td>\n",
       "      <td>4938</td>\n",
       "      <td>5409</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>eGFP:c.681_686del</td>\n",
       "      <td>((228, 'd'), (229, 'd'))</td>\n",
       "      <td>708</td>\n",
       "      <td>4120</td>\n",
       "      <td>275</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>eGFP:c.681_683del</td>\n",
       "      <td>((228, 'd'),)</td>\n",
       "      <td>988</td>\n",
       "      <td>3376</td>\n",
       "      <td>852</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eGFP:c.678_686del</td>\n",
       "      <td>((227, 'd'), (228, 'd'), (229, 'd'))</td>\n",
       "      <td>276</td>\n",
       "      <td>3050</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>eGFP:c.692_700del</td>\n",
       "      <td>((232, 'd'), (233, 'd'), (234, 'd'))</td>\n",
       "      <td>169</td>\n",
       "      <td>1978</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>eGFP:c.678_683del</td>\n",
       "      <td>((227, 'd'), (228, 'd'))</td>\n",
       "      <td>178</td>\n",
       "      <td>1140</td>\n",
       "      <td>625</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>eGFP:c.462_467del</td>\n",
       "      <td>((155, 'd'), (156, 'd'))</td>\n",
       "      <td>171</td>\n",
       "      <td>1095</td>\n",
       "      <td>702</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>eGFP:c.574_576del</td>\n",
       "      <td>((192, 'd'),)</td>\n",
       "      <td>216</td>\n",
       "      <td>1014</td>\n",
       "      <td>173</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>eGFP:c.690_692del</td>\n",
       "      <td>((231, 'd'),)</td>\n",
       "      <td>184</td>\n",
       "      <td>986</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>eGFP:c.674_679del</td>\n",
       "      <td>((226, 'd'), (227, 'd'))</td>\n",
       "      <td>907</td>\n",
       "      <td>886</td>\n",
       "      <td>9863</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>eGFP:c.690_695del</td>\n",
       "      <td>((231, 'd'), (232, 'd'))</td>\n",
       "      <td>111</td>\n",
       "      <td>855</td>\n",
       "      <td>263</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>eGFP:c.687_695del</td>\n",
       "      <td>((230, 'd'), (231, 'd'), (232, 'd'))</td>\n",
       "      <td>86</td>\n",
       "      <td>821</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>eGFP:c.690_698del</td>\n",
       "      <td>((231, 'd'), (232, 'd'), (233, 'd'))</td>\n",
       "      <td>53</td>\n",
       "      <td>756</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>eGFP:c.680_685del</td>\n",
       "      <td>((227, 's', 'V'), (228, 'd'), (229, 'd'))</td>\n",
       "      <td>109</td>\n",
       "      <td>751</td>\n",
       "      <td>240</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             DNA_hgvs                                    Protein  Baseline  \\\n",
       "0   eGFP:c.674_676del                              ((226, 'd'),)      2105   \n",
       "1   eGFP:c.458_463del  ((153, 's', 'N'), (154, 'd'), (155, 'd'))      1247   \n",
       "2   eGFP:c.681_686del                   ((228, 'd'), (229, 'd'))       708   \n",
       "3   eGFP:c.681_683del                              ((228, 'd'),)       988   \n",
       "4   eGFP:c.678_686del       ((227, 'd'), (228, 'd'), (229, 'd'))       276   \n",
       "5   eGFP:c.692_700del       ((232, 'd'), (233, 'd'), (234, 'd'))       169   \n",
       "6   eGFP:c.678_683del                   ((227, 'd'), (228, 'd'))       178   \n",
       "7   eGFP:c.462_467del                   ((155, 'd'), (156, 'd'))       171   \n",
       "8   eGFP:c.574_576del                              ((192, 'd'),)       216   \n",
       "9   eGFP:c.690_692del                              ((231, 'd'),)       184   \n",
       "10  eGFP:c.674_679del                   ((226, 'd'), (227, 'd'))       907   \n",
       "11  eGFP:c.690_695del                   ((231, 'd'), (232, 'd'))       111   \n",
       "12  eGFP:c.687_695del       ((230, 'd'), (231, 'd'), (232, 'd'))        86   \n",
       "13  eGFP:c.690_698del       ((231, 'd'), (232, 'd'), (233, 'd'))        53   \n",
       "14  eGFP:c.680_685del  ((227, 's', 'V'), (228, 'd'), (229, 'd'))       109   \n",
       "\n",
       "    High  Medium  Low  \n",
       "0   9468     340   13  \n",
       "1   4938    5409    9  \n",
       "2   4120     275    0  \n",
       "3   3376     852    7  \n",
       "4   3050       1    3  \n",
       "5   1978       1    0  \n",
       "6   1140     625    1  \n",
       "7   1095     702    2  \n",
       "8   1014     173    0  \n",
       "9    986      73    0  \n",
       "10   886    9863   32  \n",
       "11   855     263    0  \n",
       "12   821       0    0  \n",
       "13   756       0    0  \n",
       "14   751     240    0  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find the top 15 H mutations\n",
    "\n",
    "def top_variant_distribution(gfp_cnt, bg, fraction, filename):\n",
    "\n",
    "    top = gfp_cnt[bg][fraction].most_common(100)\n",
    "    with open(filename, 'w') as f:\n",
    "        csv_writer = csv.writer(f, delimiter=',')\n",
    "        csv_writer.writerow(['DNA_hgvs', 'Protein', 'Baseline', 'High', 'Medium', 'Low'])\n",
    "        for pair in top:\n",
    "            dna_hgvs = pair[0]\n",
    "            try:\n",
    "                protein = gfp_translate[bg][dna_hgvs]\n",
    "            except KeyError:\n",
    "                print(dna_hgvs)\n",
    "                raise\n",
    "            csv_writer.writerow([dna_hgvs, protein, gfp_cnt[bg]['baseline'][dna_hgvs], \n",
    "                                 gfp_cnt[bg]['H'][dna_hgvs], gfp_cnt[bg]['MM'][dna_hgvs], gfp_cnt[bg]['L'][dna_hgvs]] )\n",
    "\n",
    "top_variant_distribution(gfp_cnt, 'eGFP', 'H', 'Top_variants_eGFP_H.csv')\n",
    "top_variant_distribution(gfp_cnt, 'GFP8', 'H', 'Top_variants_GFP8_H.csv')\n",
    "\n",
    "df_top = pd.read_csv('Top_variants_GFP8_H.csv')\n",
    "df_top = pd.read_csv('Top_variants_eGFP_H.csv')\n",
    "df_top.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "()"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gfp_translate['eGFP']['eGFP:c.39A>T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
